### ISSUE TITLE
Deleted or non selected source files are linked in

### ISSUE THREAD

**Issue by fjmolinas** [RIOT-OS/RIOT#16942]
#### Description

This might also occur in other cases but for example when a `MODULE` has different implementations which are selected vía a `PSEUDOMODULE` or when `SRC` file selection changes the old files are still linked in which leads to either bloated code to compilation errors. This was introduced with  #14754 where object files started to be linked instead on relying on archives, archives were always recreated so this was not an issue before #14754 https://github.com/RIOT-OS/RIOT/blob/c1ad43566855c9f95b3743d9a9ce28a1d8efe6f3/Makefile.base#L95-L98

#### Steps to reproduce the issue

This can be done in any case where SRC or PSEUDOMODULES are selected/deselected. An example that worked at the time of #14754:

```
make -C tests/pkg_openwsn/ all -j7
```

```
SEMODULE=openwsn_sctimer_ztimer make -C tests/pkg_openwsn/ all -j7
```

#### Expected results

Only new files are linked in.

#### Actual results

Old and new files are linked in.

```
/home/francisco/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_ztimer.o: in function `sctimer_init':
/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_ztimer.c:80: multiple definition of `sctimer_init'; /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_rtt.o:/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_rtt.c:84: first defined here
/home/francisco/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_ztimer.o: in function `sctimer_set_callback':
/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_ztimer.c:116: multiple definition of `sctimer_set_callback'; /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_rtt.o:/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_rtt.c:95: first defined here
/home/francisco/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_ztimer.o: in function `sctimer_setCompare':
/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_ztimer.c:120: multiple definition of `sctimer_setCompare'; /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_rtt.o:/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_rtt.c:119: first defined here
/home/francisco/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_ztimer.o: in function `sctimer_readCounter':
/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_ztimer.c:149: multiple definition of `sctimer_readCounter'; /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_rtt.o:/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_rtt.c:150: first defined here
/home/francisco/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_ztimer.o: in function `sctimer_enable':
/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_ztimer.c:158: multiple definition of `sctimer_enable'; /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_rtt.o:/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_rtt.c:163: first defined here
/home/francisco/opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_ztimer.o: in function `sctimer_disable':
sctimer_ztimer.c:(.text.sctimer_disable+0x0): multiple definition of `sctimer_disable'; /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/openwsn/sctimer_rtt.o:/home/francisco/workspace/RIOT/pkg/openwsn/contrib/sctimer_rtt.c:168: first defined here
collect2: error: ld returned 1 exit status
make: *** [/home/francisco/workspace/RIOT/tests/pkg_openwsn/../../Makefile.include:585: /home/francisco/workspace/RIOT/tests/pkg_openwsn/bin/iotlab-m3/tests_pkg_openwsn.elf] Error 1
```

#### Versions

2021.10

**Comment by fjmolinas** [comment]
Note that of course `clean` fixes the issue but IMO that should not be needed.

**Comment by fjmolinas** [comment]
Fixed by #16945
